File: 45. Jump Game II.cpp
class Solution {
public:
    int jump(vector<int>& nums) {
        return helper(nums, 0, 0, 0);
    }

    int helper(vector<int>& nums, int idx, int threshold, int accum) {
        if (idx == nums.size() - 1)
            return 0;
        else if (nums[idx] + idx >= nums.size() - 1)
            return accum + 1;
            threshold++;
            int maxIdx = threshold, maxDist = nums[threshold] + threshold, upper = idx + nums[idx];
            for (int i = ++threshold; i <= upper; i++) {
                if (nums[i] + i > maxDist) {
                    maxIdx = i;
                    maxDist = i + nums[i];
                }
            }
        return helper(nums, maxIdx, upper, accum + 1);
    }
};

File: java/45. Jump Game II.java
// O(n)
class Solution {
    public int jump(int[] nums) {
        if (nums.length <= 1) {
            return 0;
        }
        int jumps = 0;
        int minIdx = -1, maxIdx = 0, curIdx = 0, nextIdx = 0;
        while (curIdx > minIdx) {
            if (curIdx + nums[curIdx] >= nums.length - 1) {
                return jumps + 1;
            }
            nextIdx = Math.max(nextIdx, curIdx + nums[curIdx]);
            --curIdx;
            if (curIdx == minIdx) {
                minIdx = maxIdx;
                maxIdx = nextIdx;
                curIdx = nextIdx;
                ++jumps;
            }
        }
        return jumps;
    }
}
